src/embedded_flash/syscall.rs

```diff
@@ -157,83 +192,127 @@ impl Storage for SyscallStorage {
+        };
+        if !locations.page_size.is_power_of_two() {
             return Err(StorageError::CustomError);
         }
-        Ok(())
+        for i in 0..memop(memop_nr::STORAGE_CNT, 0)? {
+            let storage_type = memop(memop_nr::STORAGE_TYPE, i)?;
+            match storage_type {
+                storage_type::PARTITION | storage_type::METADATA => (),
+                _ => continue,
+            };
+            let storage_ptr = memop(memop_nr::STORAGE_PTR, i)?;
+            let storage_len = memop(memop_nr::STORAGE_LEN, i)?;
+            if !locations.is_page_aligned(storage_ptr) || !locations.is_page_aligned(storage_len) {
+                return Err(StorageError::CustomError);
+            }
+            let range = ModRange::new(storage_ptr, storage_len);
+            match storage_type {
+                storage_type::PARTITION => {
+                    locations.partition = locations
+                        .partition
+                        .append(range)
+                        .ok_or(StorageError::NotAligned)?
+                }
+                storage_type::METADATA => {
+                    locations.metadata = locations
+                        .metadata
+                        .append(range)
+                        .ok_or(StorageError::NotAligned)?
+                }
+                _ => (),
+            };
+        }
+        if locations.partition.is_empty() || locations.metadata.is_empty() {
+            Err(StorageError::CustomError)
+        } else {
+            Ok(locations)
+        }
+    }
+
+    fn is_page_aligned(&self, x: usize) -> bool {
+        is_aligned(self.page_size, x)
     }
 }
 
-fn find_slice<'a>(
-    slices: &'a [&'a [u8]],
-    mut start: usize,
-    length: usize,
-) -> StorageResult<&'a [u8]> {
-    for slice in slices {
-        if start >= slice.len() {
-            start -= slice.len();
-            continue;
+impl UpgradeStorage for SyscallUpgradeStorage {
+    fn read_partition(&self, offset: usize, length: usize) -> StorageResult<&[u8]> {
+        if self
+            .partition
+            .contains_range(&ModRange::new(offset, length))
+        {
+            Ok(unsafe { core::slice::from_raw_parts(offset as *mut u8, length) })
+        } else {
+            Err(StorageError::OutOfBounds)
         }
-        if start + length > slice.len() {
-            break;
+    }
+
+    fn write_partition(&mut self, offset: usize, data: &[u8]) -> StorageResult<()> {
+        if self
+            .partition
+            .contains_range(&ModRange::new(offset, data.len()))
+        {
+            // Erases all pages that have their first byte in the writte range.
+            // Since we expect calls in order, we don't want to erase half-written pages.
+            for address in ModRange::new(offset, data.len()).aligned_iter(self.page_size) {
+                erase_page(address, self.page_size)?;
+            }
+            write_slice(offset, data)
```

Comment:
So this is not where we check that we only write on top of `0xff`s. Is it higher in the call stack?

Otherwise we could check `read_partition(offset, data.len())?.iter().all(|x| x == 0xff)`.

Reply:
We do check higher up in the call stack by reading back the slice and checking if it matches our expectation of what was writing. But adding the check here allows to forward-fix and special case the first page to erase and re-write. In case something case wrong between the erase and rewrite of old data, it is kind of unexpected behaviour from an outside look, but it might make sense if we expect the caller to do something like:

- Try to greedy write slice by slice.
- If a slice write fails, retry the same slice.
- If the retry also fails, restart the whole process.

So how about I add the `all(|x| x == 0xff)`, and if it fails,
- read the first page
- erase the first page
- write the saved page until `offset`
- ?

Reply:
I'm not sure it's worth the complexity. I think for embedded it's ok to say: first failure -> fail the process -> let the user retry or  debug.

If we check higher-up that we read what we wrote, then all good.

Reply:
Okay, then I'll keep this write simple for now and add features if we notice usability issues or change requirements.